linker.ts

```diff
@@ -1,72 +1,133 @@
 import assert from 'assert';
 import { keccak256 } from 'js-sha3';
+import { isNil, isObject } from './common/helpers';
 
-function libraryHashPlaceholder (input) {
-  return '$' + keccak256(input).slice(0, 34) + '$';
+/**
+ * Generates a new-style library placeholder from a fully-qualified library name.
+ *
+ * Newer versions of the compiler use hashed names instead of just truncating the name
+ * before putting it in a placeholder.
+ *
+ * @param fullyQualifiedLibraryName
+ */
+function libraryHashPlaceholder (fullyQualifiedLibraryName) {
+  return `$${keccak256(fullyQualifiedLibraryName).slice(0, 34)}$`;
 }
 
-const linkBytecode = function (bytecode, libraries) {
+/**
+ * Used to support legacy library names and the current hash library names,
+ * replaces the names with the provided hex address.
+ *
+ * @param bytecode
+ * @param address
+ * @param label
+ */
+function replacePlaceholder (bytecode, label, address) {
```

Comment:
```suggestion
/**
 * Finds all placeholders corresponding to the specified library label and replaces them
 * with a concrete address. Works with both hex-encoded and binary bytecode as long as
 * the address is in the same format.
 *
 * @param bytecode Bytecode string.
 * @param label Library label, either old- or new-style. Must exactly match the part between `__` markers in the 
 *     placeholders. Will be padded with `_` characters if too short or truncated if too long.
 * @param address Address to replace placeholders with. Must be the right length.
 *     It will **not** be padded with zeros if too short.
  */
function replacePlaceholder (bytecode, label, address) {
```